### ISSUE TITLE
import meta graph bug reading model with scope name changed

### ISSUE THREAD

**Issue by chenghuige** [tensorflow/tensorflow#4757]
Same issue for tf 0.10 and 0.11.rc0.
The problem is if you read one model, change the top scope name, write a new model, then you read the new model, but still you got the old top scope name.
How to reproduce 

- create model
python create-model.py

        import tensorflow as tf  

        sess = tf.InteractiveSession()  

        with tf.variable_scope('old'):  

            w = tf.get_variable('w', shape=[1], initializer=tf.constant_initializer(1.0))  

        sess.run(tf.initialize_all_variables())  

        tf.train.Saver().save(sess, '/tmp/old.model')  

- read this model (using import meta graph), then rename top scope and save to one new model
python rename-scope.py
old_vars: [u'old/w:0']
new_vars: [u'new/w']  

        import tensorflow as tf  

        sess = tf.InteractiveSession()  

        saver = tf.train.import_meta_graph('/tmp/old.model.meta')  

        saver.restore(sess, '/tmp/old.model')  

        src_vars = [v for v in tf.all_variables() if v.name.startswith('old')]  

        print('old_vars:', [v.name for v in src_vars])

        out_vars = {v.name[:v.name.rfind(':')].replace('old', 'new', 1): v for v in src_vars}  

        print('new_vars:', [key for key in out_vars])  

        tf.train.Saver(var_list=out_vars).save(sess, '/tmp/new.model') 
 
 
- read new model(It is ok if you buid graph from scratch with new scope, but not ok if using import meta graph again)
#-------------------this is ok
python read-renamed-buildgraph.py  
tf.all_variables: [u'new/w:0']
w val: [ 1.]

        import tensorflow as tf  

        sess = tf.InteractiveSession()

        with tf.variable_scope('new'):
            w = tf.get_variable('w', shape=[1], initializer=tf.constant_initializer(2.0))  

        tf.train.Saver().restore(sess, '/tmp/new.model')  

        print('tf.all_variables:', [v.name for v in tf.all_variables()])  

        print('w val:',  w.eval())

  #---------------this is wrong 
  python read-renamed-metagraph.py
  tf.all_variables: [u'old/w:0']

       import tensorflow as tf  

       sess = tf.InteractiveSession()  

       saver = tf.train.import_meta_graph('/tmp/new.model.meta')

       saver.restore(sess, '/tmp/new.model')

       print('tf.all_variables:', [v.name for v in tf.all_variables()])

**Comment by tatatodd** [comment]
@chenghuige The problem is in your second step `rename-scope.py`, where you run the following two lines:
```
out_vars = {v.name[:v.name.rfind(':')].replace('old', 'new', 1): v for v in src_vars}
tf.train.Saver(var_list=out_vars).save(sess, '/tmp/new.model')
```

I think you're assuming that this step is sufficient to rename the `old/w` variable to `new/w`.  In reality it's not.  Note that the python string replace method returns a copy of the string, and doesn't mutate the string itself:
https://docs.python.org/2/library/string.html#string.replace

What are you actually trying to do?

**Comment by chenghuige** [comment]
@tatatodd  I thins string replace here is fine, you can see my example code of   
python read-renamed-buildgraph.py it actually works if not reading using meta graph.

[saving-and-restoring](https://www.tensorflow.org/versions/r0.11/how_tos/variables/index.html#saving-and-restoring)
# Create some variables.
v1 = tf.Variable(..., name="v1")
v2 = tf.Variable(..., name="v2")
...
# Add ops to save and restore only 'v2' using the name "my_v2"
saver = tf.train.Saver({"my_v2": v2})
# Use the saver object normally after that.
...

Actually by doing this I can load two same models(same model just differnt saving training steps, like bow model of step 1 and step 100) into the same session.
I assume by doing this I save variable using new name. So to avoid two load two models with same weight.

**Comment by tatatodd** [comment]
@chenghuige I'm sorry, my comment about python `string.replace` was irrelevant; I'd misinterpreted your code.

But the problem still remains in your second step `rename-scope.py`.  After that step, the checkpoint `/tmp/new.model` will have a single entry for `new/w`.  That is what you want.  But the MetaGraphDef `/tmp/new.model.meta` will still hold a graph that contains the `old/w` node.  That is because the `out_vars` that you passed to `tf.train.Saver` only renamed the checkpoint entry, but not the graph itself.

There is no bug in `import_meta_graph`, since it's just loading the graph that was provided.

We could try to make the argument that there's a bug in `tf.train.Saver`.  I.e. we passed it new names for some variables, so maybe it should rename the graph nodes too.  But that doesn't really work; e.g. the `old/w` variable also has related nodes `old/w/Initializer/Const`, `old/w/Assign`, `old/w/read` that are used to initialize the variable.  In general isn't possible for us to infer the renames that would be required.

---

So basically the best way to accomplish your goal is your step 3 `read-renamed-buildgraph.py`.

Hope that helps!  If you have further questions, just comment on this issue, and I'll re-open.

**Comment by chenghuige** [comment]
Very clear explanation, thanks! @tatatodd
